### ISSUE TITLE
Full line comment can cause G30 to execute twice

### ISSUE THREAD

**Issue by gloomyandy** [chrishamm/DuetSoftwareFramework#141]
Hi Chris,
I think there may be a problem when a file contains a full line comment just before a G30 command (and possibly other commands that trigger a macro execution). So for instance the following sequence:
```
G28
M561 ; 1 clear any bed transform
; G28
G30 P0 X40 Y145 Z-99999 ; 1
G30 P1 X140 Y145 Z-99999 ; 2
```

Can result in the line:
```
G30 P0 X40 Y145 Z-99999 ; 1
```
being executed twice. I've enabled the firmware gcode debug and seen the command issued twice. I have a log file from DCS (which is rather large), you can find it here: https://drive.google.com/open?id=1bZkc8LGtHFNhTTjvBubaSQx3c2X9pXYL

The above test code was saved as a file called testcomment.gcode and executed.

I'm not 100% sure but I think that what is happening is...
1 After the G28 is complete all four remaining lines get buffered and sent to the firmware.
2. The M561 commend is executed and DSF receives a BinaryCodeReplyFlag for this.
3. HandleReply is called for M561 and this removes M561 from BufferedCodes, so at this point the "; G28" is in BufferCodes[0].
4. The firmware skips the comment and does not generate any sort of reply and starts to execute the G30
5. The G30 command generates macro calls to deploy and retract the probe (in my case those files do not exist).
6. The "; G28" command is removed from BufferedCodes (as the startCode) and the stack is pushed, leaving the (currently actually executing) "G30 P0" and "G30 P1" command in the SuspendedCodes list.
7. The retract macro gets called and various pop/push operations happen.
8. Eventually the firmware completes the "G30 P0" command and HandleReply is called, this pops the stack and reports: ""==> Finished starting code: ; G28". I'm pretty sure a comment should never be the starting code for a macro call!
9. Now things get sort of back to normal and the SuspendedCodes are retsarted, but note that in this case this includes the "G30 P0" command that has already been executed once by the firmware 
10. The second copy of "G30 P0" executes pretty much as normal and then things continue until the file is complete.

So basically I think the problem is that the full line comment "; G28" is being left in BufferedCodes when it should have been removed. Perhaps an additional call to ResolveCommentCodes is needed either at the start of DoMacro or Push? I'm not really sure about the best way to handle it.

Let me know if you need any further information.

**Comment by chrishamm** [comment]
Thanks, just in time! I think your assumption is quite close. The code reply handler in DCS already takes care of this, but the macro handler does not. I've just committed a hotfix for the upcoming 3.1.0 package which should resolve this.
